File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.py
import sys
input = sys.stdin.readline

while True:

    h_list=list(map(int,input().split()))
    if len(h_list)==1:
        break
    h_list = h_list[1:]
    result = 0
    stack=[]
    for i,h in enumerate(h_list+[0]):
        # print(i,h)
        tmp=i
        while True:
            if not stack or stack[-1][0]<=h:
                stack.append([h,tmp])
                break
            else:
                h2,j = stack.pop()
                # print('pop', h2*(i-j))
                tmp=j
                result = max(result, h2*(i-j))
        # print('ss',stack,result)

    print(result)

File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();
        
        while (true) {
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            if (n == 0) break;
            
            int[] hList = new int[n + 1];
            for (int i = 0; i < n; i++) {
                hList[i] = Integer.parseInt(st.nextToken());
            }
            
            long result = 0;
            Stack<int[]> stack = new Stack<>();
            
            for (int i = 0; i <= n; i++) {
                int h = (i == n) ? 0 : hList[i];
                int tmp = i;
                
                while (!stack.isEmpty() && stack.peek()[0] > h) {
                    int[] top = stack.pop();
                    tmp = top[1];
                    result = Math.max(result, (long) top[0] * (i - tmp));
                }
                stack.push(new int[]{h, tmp});
            }
            sb.append(result).append("\n");
        }
        System.out.print(sb);
    }
}